File: 539.minimum-time-difference.py
#
# @lc app=leetcode id=539 lang=python3
#
# [539] Minimum Time Difference
#

# @lc code=start
class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        time_seen = [False]*60*24
        for time in timePoints:
            h = int(time[0:2])
            m = int(time[3:5])
            if time_seen[60*h+m]:
                return 0
            else:
                time_seen[60*h+m] = True
        pos_first = 0
        while not time_seen[pos_first]:
            pos_first += 1
        pos_last = 60*24-1
        while not time_seen[pos_last]:
            pos_last -= 1
        res = pos_first - pos_last + 60 * 24
        if res == 1:
            return 1
        count = 1
        print(m)
        for i in range(pos_first+1, pos_last+1):
            if time_seen[i]:
                res = min(res, count)
                if res == 1:
                    return 1
                count = 1
            else:
                count += 1
        return res


# @lc code=end

File: java/539.minimum-time-difference.java
/*
 * @lc app=leetcode id=539 lang=java
 *
 * [539] Minimum Time Difference
 */

// @lc code=start
class Solution {
    public int findMinDifference(List<String> timePoints) {
        boolean[] time_seen = new boolean[60 * 24];
        for (String time : timePoints) {
            int h = Integer.parseInt(time.substring(0, 2));
            int m = Integer.parseInt(time.substring(3, 5));
            if (time_seen[60 * h + m]) {
                return 0;
            } else {
                time_seen[60 * h + m] = true;
            }
        }
        int pos_first = 0;
        while (!time_seen[pos_first]) {
            ++pos_first;
        }
        int pos_last = 60 * 24 - 1;
        while (!time_seen[pos_last]) {
            --pos_last;
        }
        int res = pos_first - pos_last + 60 * 24;
        if (res == 1) {
            return 1;
        }
        int count = 1;
        for (int i = pos_first + 1; i != pos_last + 1; ++i) {
            if (time_seen[i]) {
                res = Math.min(count, res);
                if (res == 1) {
                    return 1;
                }
                count = 1;
            } else {
                ++count;
            }
        }
        return res;
    }
}
// @lc code=end
